internal/charmstore/store.go

```diff
@@ -523,23 +503,23 @@ func (s *Store) FindBestEntity(url *charm.URL, channel Channel, fields map[strin
 			return nil, errgo.Mask(err)
 		}
 		// If a channel was spedified make sure the entity is in that channel.
-		if channel == StableChannel && !entity.Stable {
+		if channel == mongodoc.StableChannel && !entity.Stable {
 			return nil, errgo.WithCausef(nil, params.ErrNotFound, "no matching charm or bundle for %s", url)
 		}
-		if channel == DevelopmentChannel && !entity.Development {
+		if channel == mongodoc.DevelopmentChannel && !entity.Development {
 			return nil, errgo.WithCausef(nil, params.ErrNotFound, "no matching charm or bundle for %s", url)
 		}
 		return entity, nil
 	}
+
 	switch channel {
-	case StableChannel, NoChannel:
-		return s.findStableEntity(url, fields)
-	case DevelopmentChannel:
-		return s.findDevelopmentEntity(url, fields)
-	case UnpublishedChannel:
+	case mongodoc.UnpublishedChannel:
 		return s.findUnpublishedEntity(url, fields)
```

Comment:
I'm not sure. I can't think of a way of updating ChannelEntities on initial publish in a non-racy way so that we guarantee that looking up an unversioned entity in the Unpublished channel will always return the latest unpublished entity.